### PR TITLE
[Dynamo] Guard source-backed identity comparisons

### LINKED ISSUE #175267: [dynamo] Audit `x is y` code path
### 🐛 Describe the bug

While working on https://github.com/pytorch/pytorch/pull/175266, Dynamo was failing `test_getset_descriptor_objclass_identity` added test.

The fix was 

```
        elif name in ("__objclass__", "__name__"):
            source = self.source and AttrSource(self.source, name)
            return VariableTracker.build(tx, getattr(self.desc, name), source)
```

but even if this fix was not there, I was not expecting it fail. After debugging, I found the issue is in `handle_is` call, where earlier we were doing 

Image

After replacing it with 

Image

the test started passing without even the special handling of `__objclass__`. 

This exposes a concern that `is` is not working properly in Dynamo. This issue is to audit and perhaps ask LLMs to write many complex `is` tests to expose issues.

### Error logs

_No response_

### Versions

NA

cc @chauhang @penguinwu

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #184830

Dynamo's identity helper only trusted VariableTracker real Python values. Source-backed values such as aliases, inherited descriptors, and class-dict descriptor entries can represent concrete Python objects while reporting NO_SUCH_SUBOBJ, so `is` comparisons could be folded to false even when eager identity was true. Tensor aliases also exposed a stale-cache path when a global tensor was canonicalized with an input without preserving the alias dependency.

This change resolves only stable source-backed identity values through the current frame and installs ID guards before folding the result. The stable set is intentionally narrow: direct local/global/getitem sources, stable class attributes, class-dict descriptor entries whose class lookup returns the same object, and selected descriptor metadata such as `__func__`. Comparisons that cannot be proven and guarded now graph-break instead of falling through to unsafe constant folding. `UserMethodVariable` also stops reporting the underlying function as the bound method's real object so fresh method identities are not misrepresented.

For tensor identity, the existing fake-example identity path is preserved for sourceless tensor aliases like `x.contiguous() is x`, while duplicate-input guards now cover aliases across local/global/nonlocal sources so rebinding a captured tensor source invalidates the cache. I considered broadly resolving arbitrary sources, but descriptor lookups can synthesize fresh objects, so the narrower guarded-source approach avoids immediate guard failures and stale folds.

Fixes #175267
Generated by my agent

Test Plan:
- python test/dynamo/test_repros.py ReproTestsDeviceCUDA.test_is_with_mutated_tensor_alias_source_cuda
- python test/dynamo/test_repros.py ReproTestsDeviceCUDA.test_operator_is_with_source_aliases_cuda ReproTestsDeviceCUDA.test_inherited_c_descriptor_identity_cuda ReproTestsDeviceCUDA.test_is_with_mutated_class_attr_tensor_cuda ReproTestsDeviceCUDA.test_is_with_tensor_result_alias_cuda ReproTestsDeviceCUDA.test_is_with_mutated_tensor_alias_source_cuda ReproTestsDeviceCUDA.test_is_with_unsafe_instance_attr_source_cuda ReproTestsDeviceCUDA.test_is_with_descriptor_sources_from_class_dict_cuda ReproTestsDeviceCUDA.test_is_with_mutated_staticmethod_source_cuda ReproTestsDeviceCUDA.test_is_with_fresh_bound_method_source_cuda ReproTestsDeviceCUDA.test_is_with_mutated_classmethod_sources_cuda ReproTestsDeviceCUDA.test_getset_descriptor_objclass_identity_cuda ReproTestsDeviceCUDA.test_inspect_signature_callable_class_cuda
- python - <<'PY' ... with torch._dynamo.config.patch(use_lamba_guard_for_object_aliasing=False) for the module-global tensor alias/rebind repro
- lintrunner -a

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98 @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I fixed the unclassified Dynamo CI failures by extending guarded identity comparison handling for TypeSource values, dataclass slot-backed attributes, fresh C-level bound method lookups, and sourceless container comparisons observed in the HuggingFace Whisper path. Local targeted Dynamo tests, lintrunner, and the GPT2/Whisper HuggingFace graph-break probes now pass.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I fixed the unclassified ASAN export failure in test_dynamic_lstm_training_ir_to_decomp_strict. The root cause was that weakref variables only exposed their ref() source, so ref is not None graph-broke; the update preserves the weakref object source and guards it before folding weakref-vs-None identity.

**Comment by jansel** [comment]
My agent says I rebased the PR onto current main and fixed the two remaining unclassified CI regressions. The update preserves the UserMethodVariable identity fix while making bound method __dict__ read from the underlying function dict, and it lets guarded builtin/singleton identity checks such as TreeSpec.type is dict fold safely.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I fixed the unclassified Dynamo-wrapped custom-ops CI failure by making _safe_type_repr tolerate types.GenericAlias values while preserving the custom-metaclass repr bypass for real classes. I validated with the targeted guard regression, the failing custom-ops test under PYTORCH_TEST_WITH_DYNAMO=1, pytest's selector for that test, and lintrunner -a.

**Comment by jansel** [comment]
My agent says I rebased onto current main and fixed the new Dynamo identity CI regressions. The follow-up avoids tensor ID_MATCH guards for tensor-vs-non-tensor false identity checks, allows guarded enum/class identity values, and handles stable builtin function objects like enum _new_member_ versus object.__new__ before the fresh bound-method path. I validated the failing reduce/deepcopy/FSDP selectors, added focused regressions, reran the identity test set, and ran lintrunner -a.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased onto current main and fixed the remaining dynamic-shapes module comparison CI regression. The issue was that object equality for unspecialized nn.Module values falls back to identity, but NNModuleSource wrappers were not included in the guarded source-backed identity resolver, so layer == self.layer0 graph-broke. The update resolves and guards NNModuleSource identities through the same narrow source-backed path, and the reproduced Dynamo/dynamic-shapes/custom-ops selectors plus lintrunner now pass locally.

**Comment by jansel** [comment]
My agent says I rebased onto current main and fixed the remaining PR-related Dynamo identity failures. The update resolves guarded closure-cell sentinel sources for nested-call wrappers, stops treating non-None constants like the None fast path, and graph-breaks SymNode identity comparisons instead of compiling stale Python object identity results. I validated the failed nested graph-break selector, SymNode identity regressions, the representative DRCI selectors, the broader identity repro set, and lintrunner -a.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased onto current main and fixed the remaining PR-related Dynamo identity CI failures. The SymNode identity handling now still graph-breaks unsafe symbolic scalar identity cases, but routes safe SymNode-vs-nonconstant comparisons and type-disjoint identity checks through guarded identity reasoning, which restores the upstream , dynamic , and optree namespace cases. I validated the DRCI representatives, nested graph-break wrappers, custom-ops, guard-manager, deepcopy, tree-map, and lintrunner locally.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says clarification on the previous CI comment: the restored safe cases are SymNode is-not list, dynamic copy.deepcopy(dict), and optree namespace identity checks. Unsafe symbolic scalar identity comparisons still graph-break.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased onto current main and fixed the remaining CPython enum identity graph break. The issue was that builtin classmethod descriptors such as object.__init_subclass__ create fresh bound builtin method objects on lookup, so cls.__init_subclass__ is not object.__init_subclass__ should fold false/true only through the same guarded fresh-method path. The update recognizes that narrow builtin-class classmethod descriptor case and adds focused Dynamo regressions for Color.__init_subclass__ versus object.__init_subclass__ and repeated object.__init_subclass__ lookups. I validated the new focused selector, the prior enum/safe-repr/custom-ops representatives, and lintrunner locally; the local CPython 3.13 test file cannot run in this Python 3.12 environment.

**Comment by jansel** [comment]
My agent says I rebased the PR onto current main and addressed the latest DRCI enum cleanup: the CPython 3.13 enum test now unexpectedly succeeds, so I removed the stale Dynamo expected-failure marker. I also revalidated the focused enum repro, the safe type repr guard test, the Dynamo-wrapped custom-ops representative before the final rebase, and lintrunner -a.

**Comment by jansel** [comment]
My agent says I rebased the PR onto current main and fixed the remaining Dynamo identity regressions from DRCI. Source-backed identity resolution now covers set elements and const-dict keys, and speculative source evaluation falls back to unknown instead of raising when an imported-module source cannot be resolved in the root frame. This addresses the local reproductions in TestSetGuards.test_set_recompile_on_key_change, MiscTestsPyTree.test_pytree_tree_flatten_unflatten_cxx, and the Dynamo-wrapped LinearCrossEntropyOptions test. I also revalidated the enum/safe_type_repr/custom_ops/decorators/compiled-autograd representatives, the identity repro sweep, and lintrunner -a.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased onto current main and fixed the current PR-related CI failures. The update makes _safe_type_repr use the object's real metaclass before calling type.__repr__, covering GenericAlias and spoofed __class__ cases; skips the _new_member_ repro on Python versions where Enum does not expose that helper; and removes the stale CPython 3.13 enum expected-failure marker. I validated the failed guard-manager, enum, dynamic-shapes, and Dynamo-wrapped custom-ops representatives plus lintrunner -a.

**Comment by jansel** [comment]
My agent says I rebased onto current main and addressed the latest CI signals. The CPython 3.13 enum failure was a stale expected-failure marker for the now-passing PlainFlag enum_in_enum_out cases, so those markers are removed. The inductor_timm mobilenetv2_100 failure occurred before Dynamo ran because two eager CUDA TIMM training runs differed at exact tolerance in near-zero BatchNorm gradients, so the benchmark validators now scope that known flake to CUDA TIMM training baselines instead of treating it as a PR regression. I validated the targeted Dynamo identity representatives, safe_type_repr, the validator behavior, and lintrunner -a.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased onto current main and fixed the latest DRCI recursive deepcopy failure. The identity path now handles sourceless newly-created user objects whose class source is known: bound C methods such as object.__reduce_ex__ are folded non-None only after guarding both the class attribute and MRO descriptor sources and rejecting pending or actual instance shadowing. I added coverage for recursive deepcopy, class shadowing invalidation, and instance shadowing, and revalidated the focused bound-builtin identity tests plus lintrunner -a.

**Comment by jansel** [comment]
My agent says I rebased onto current main and fixed the two current DRCI new failures. The identity path now handles stable metaclass property sources such as GroupMember.WORLD, and it can prove bound builtin methods such as object.__reduce_ex__ are non-None for objects/classes created during tracing when the static descriptor still matches. I added focused regressions for both the fake distributed subgroup path and recursive deepcopy of a newly-created class, and revalidated the broader Dynamo identity repro set plus lintrunner -a.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I fixed the current DRCI new failure in dynamo-cpython-test: test/cpython/v3_13/test_enum.py::TestIntEnumClass::test_enum_in_enum_out was an unexpected success after the Dynamo identity fix. The update removes the remaining generated CPython 3.13 test_enum_in_enum_out expected-failure markers, not just the earlier Plain enum variants. I validated the focused Dynamo enum identity regression and lintrunner locally; the exact CPython 3.13 shard is not runnable in this Python 3.12 environment because enum.EnumDict is unavailable.

**Comment by jansel** [comment]
My agent says I fixed the current DRCI failure in test/cpython/v3_13/test_functools.py::TestReducePy::test_iterator_usage. The root cause was that pure-Python functools.reduce compares its omitted initial argument against an object() default sentinel, but Dynamo did not resolve DefaultsSource in guarded identity comparisons. The update adds DefaultsSource to the stable source-backed identity path and adds a regression that checks recompilation when __defaults__ changes. I also removed the earlier broad CUDA TIMM benchmark flake-scoping change from this PR because current main has a narrower eager_two_runs_differ classification. Local focused reduce harness, identity repro tests, and lintrunner -a pass.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased this PR onto current main and resolved the graph-break registry conflict by preserving both entries. The rebase exposed new identity-soundness cases around MethodWrapperVariable, custom metaclass/descriptor side effects, mutable SkipGuardSource values, and GroupMember.WORLD descendant guards; the update now uses explicit guarded/trusted source contracts and conservatively graph-breaks arbitrary runtime descriptors. The full added identity set, attribute-protocol tests, dynamic/nested variants, historical CI representatives, registry lint, and lintrunner pass locally. Fresh CI is pending.

**Comment by jansel** [comment]
My agent says I addressed the current DRCI failures and the latest Claude review. The repeated Dynamo failures came from ContextVar.get() producing a guarded source that identity comparison did not resolve; the update now handles ContextVarGetSource, buffers guards until the full proof succeeds, treats source-resolution failures as conservative graph breaks, and uses TYPE_MATCH rather than object-ID guards for type-disjoint comparisons. I also moved all device-independent identity tests to CPU coverage and removed the now-passing CPython 3.13 enum test_basics expected-failure family. The focused identity, config, distributed, compiled-autograd, dynamic/nested, full attribute-protocol, registry-lint, and lintrunner checks pass locally; fresh CI is starting.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased this PR onto current main and addressed both the latest review and the DRCI failure. Concrete identity now precedes tracker-type inference; MRO guard construction is transactional; untrusted SkipGuard aliases preserve their provenance even with reversed operands; sourceless C-method proofs require an immutable MRO prefix; and unresolved super/getattr lookups graph-break instead of folding stale values. I kept the SkipGuard same-tracker check conservative because source canonicalization can merge independently mutable locations. I also removed all inherited Python 3.13 expected-failure markers for test_contains_same_name_diff_enum_diff_values. The focused CPU, dynamic-shape, and nested-graph-break regressions, all 107 tp_getattro tests, and lintrunner pass.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
